### ISSUE TITLE
using which tokenizer for training BERT for UTF-8 Unicode languages

### ISSUE THREAD

**Issue by ReySadeghi** [huggingface/tokenizers#893]
Hey @seyyaw, @taesiri.

TLDR; This is how the byte-level BPE works. Main advantages are:
- Smaller vocabularies
- No unknown token

This is totally expected behavior. The byte-level BPE converts all the Unicode code points into multiple byte-level characters:
 1. Each Unicode code point is decomposed into bytes (1 byte for ASCII characters, and up to 4 bytes for UTF-8 Unicode code points)
 2. Each byte value gets a "visible" character assigned to it from the beginning of the Unicode table. This is especially important because there are a lot of control characters, so we can't just have a simple mapping ASCII Table character <-> byte value. So some characters get other representations, like for example the white space `U+0020` becomes `Ġ`.

The purpose is, by doing so, you end up with an initial alphabet of 256 tokens. These 256 tokens can then be merged together to represent any other token in the vocabulary. This results in smaller vocabularies, that won't ever need an "unknown" token.

_Originally posted by @n1t0 in https://github.com/huggingface/tokenizers/issues/203#issuecomment-605105611_

**Comment by ReySadeghi** [comment]
@n1t0 
Hi,
I have the same problem mentioned above for Persian language, I got from your comment that BPE tokenizer is not possible for languages with UTF-8 Unicode characters , so which one is better? word piece tokenizer?

**Comment by Narsil** [comment]
> I have the same problem mentioned above for Persian language, I got from your comment that BPE tokenizer is not possible for languages with UTF-8 Unicode characters , so which one is better? word piece tokenizer?

I think you misunderstood. BPE is very well adapted for utf-8 heavy character. To be precise it's not really `BPE` but `ByteLevel` that is important.

The `tokens` being unreadable is actually a `feature` not a `bug` (really).
The core issue is linked to how utf-8 works and control characters (which don't have a representation).
In order to make every "byte" (in the original string) printable and make everything valid utf-8 everything has to be changed around. For instance the regular space " " (byte: 32) get transferred to "Ġ". That's to enable other ascii bytes to get a printable version (like byte 2 which is not really used in regular text, and has no represnetation).

Correct me if I am wrong, but I think the core of what you want is to know, to what piece of the original string, does the token/id belong to, right ?
If that's the case `tokenizers` provides exactly what you want by using `offsets`. It is the only way to get a valid representation (consistently) of your tokens (and means something different than `tokenizer.encode(..).tokens`)

```python
from tokenizers import Tokenizer, trainers

tokenizer = Tokenizer.from_pretrained("gpt2")

string = "This is a test"
encoded = tokenizer.encode(string)

print(encoded.tokens)
# ['This', 'Ġis', 'Ġa', 'Ġtest'], would be even more unreadable on Persian

splitted = [string[start:stop] for (start, stop) in encoded.offsets]
print(splitted)
# ['This', ' is', ' a', ' test']
```

Offsets are the only to recover something meaningful consistently just because tokenization is by nature destructive, so you cannot in general recover something meaningful from a tokenized sentence on its own (here it's also unreadable as-is, there are ways to "decode" those hard to read tokens, but other issues will start creeping up)

**Comment by ReySadeghi** [comment]
> > I have the same problem mentioned above for Persian language, I got from your comment that BPE tokenizer is not possible for languages with UTF-8 Unicode characters , so which one is better? word piece tokenizer?
> 
> I think you misunderstood. BPE is very well adapted for utf-8 heavy character. To be precise it's not really `BPE` but `ByteLevel` that is important.
> 
> The `tokens` being unreadable is actually a `feature` not a `bug` (really). The core issue is linked to how utf-8 works and control characters (which don't have a representation). In order to make every "byte" (in the original string) printable and make everything valid utf-8 everything has to be changed around. For instance the regular space " " (byte: 32) get transferred to "Ġ". That's to enable other ascii bytes to get a printable version (like byte 2 which is not really used in regular text, and has no represnetation).
> 
> Correct me if I am wrong, but I think the core of what you want is to know, to what piece of the original string, does the token/id belong to, right ? If that's the case `tokenizers` provides exactly what you want by using `offsets`. It is the only way to get a valid representation (consistently) of your tokens (and means something different than `tokenizer.encode(..).tokens`)
> 
> ```python
> from tokenizers import Tokenizer, trainers
> 
> tokenizer = Tokenizer.from_pretrained("gpt2")
> 
> string = "This is a test"
> encoded = tokenizer.encode(string)
> 
> print(encoded.tokens)
> # ['This', 'Ġis', 'Ġa', 'Ġtest'], would be even more unreadable on Persian
> 
> splitted = [string[start:stop] for (start, stop) in encoded.offsets]
> print(splitted)
> # ['This', ' is', ' a', ' test']
> ```
> 
> Offsets are the only to recover something meaningful consistently just because tokenization is by nature destructive, so you cannot in general recover something meaningful from a tokenized sentence on its own (here it's also unreadable as-is, there are ways to "decode" those hard to read tokens, but other issues will start creeping up, like spaces being ignored, or too many, or u 

> ```python
> splitted = [string[start:stop] for (start, stop) in encoded.offsets]
> print(splitted)
> ```

thanks for your complete response.

**Comment by Narsil** [comment]
Hi @ReySadeghi ,

You can do:

```python
from tokenizers import Tokenizer

tokenizer = Tokenizer.from_pretrained("gpt2") # Or your model id
tokenizer.decoder.decode("ĠØ§")
tokenizer.decoder.decode("Ùħ")
```

Sorry to insist, but for future readers you should always use that for debugging purposes in order to "see" you tokenizer better. For production environments, you should have access to the original string, and offsets will be a much better match. I stress this point, because there are a lot of issues of users being confused that `tokenizer.decode(tokenizer.encode("....."))` is not what they expect.

**Comment by Narsil** [comment]
Hi @ReySadeghi ,

You can do:

```python
from tokenizers import Tokenizer

tokenizer = Tokenizer.from_pretrained("gpt2") # Or your model id
tokenizer.decoder.decode("ĠØ§")
tokenizer.decoder.decode("Ùħ")
```

Sorry to insist, but for future readers you should always use that for debugging purposes in order to "see" you tokenizer better. For production environments, you should have access to the original string, and offsets will be a much better match. I stress this point, because there are a lot of issues of users being confused that `tokenizer.decode(tokenizer.encode("....."))` is not what they expect. 

`tokenizer.decode` is a best effort, and should output something in alignment with how a model was trained, but it can be different from the original string (even very different in some cases where puncation is removed for instance)

**Comment by ReySadeghi** [comment]
> Hi @ReySadeghi ,
> 
> You can do:
> 
> ```python
> from tokenizers import Tokenizer
> 
> tokenizer = Tokenizer.from_pretrained("gpt2") # Or your model id
> tokenizer.decoder.decode("ĠØ§")
> tokenizer.decoder.decode("Ùħ")
> ```
> 
> Sorry to insist, but for future readers you should always use that for debugging purposes in order to "see" you tokenizer better. For production environments, you should have access to the original string, and offsets will be a much better match. I stress this point, because there are a lot of issues of users being confused that `tokenizer.decode(tokenizer.encode("....."))` is not what they expect.

thanks.

**Comment by Narsil** [comment]
Closing this, feel free to reopen.
